staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -0,0 +1,111 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"time"
+
+	utilclock "k8s.io/apimachinery/pkg/util/clock"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/authentication/authenticator"
+	"k8s.io/apiserver/pkg/authentication/user"
+)
+
+type cacheRecord struct {
+	user user.Info
+	ok   bool
+	err  error
+
+	expires time.Time
+}
+
+type cachedTokenAuthenticator struct {
+	authenticator authenticator.Token
+
+	successTTL time.Duration
+	failureTTL time.Duration
+
+	cache cache
+
+	clock utilclock.Clock
+}
+
+type cache interface {
+	// given a key, return the record, and whether or not it existed
+	get(string) (*cacheRecord, bool)
+	// caches the record for the key, unless the cache already contains an existing record with an expiration further in the future.
+	// returns the record for the key.
+	set(string, *cacheRecord) *cacheRecord
+	// removes the specified record for the existing key
+	remove(string, *cacheRecord)
+}
+
+// New returns a token authenticator that caches the results of the specified authenticator. A ttl of 0 bypasses the cache.
+func New(authenticator authenticator.Token, successTTL, failureTTL time.Duration) authenticator.Token {
+	return &cachedTokenAuthenticator{
+		authenticator: authenticator,
+		successTTL:    successTTL,
+		failureTTL:    failureTTL,
+		cache:         newStripedCache(25, fnvKeyFunc, newSimpleCache),
+		clock:         utilclock.RealClock{},
+	}
+}
+
+// AuthenticateToken implements authenticator.Token
+func (a *cachedTokenAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+	if record, ok := a.getCachedResult(token); ok {
+		return record.user, record.ok, record.err
+	}
+
+	user, ok, err := a.authenticator.AuthenticateToken(token)
+
+	switch {
+	case ok && a.successTTL > 0:
+		a.setCachedResult(token, &cacheRecord{user: user, ok: ok, err: err, expires: a.clock.Now().Add(a.successTTL)}, a.successTTL)
+	case !ok && a.failureTTL > 0:
+		a.setCachedResult(token, &cacheRecord{user: user, ok: ok, err: err, expires: a.clock.Now().Add(a.failureTTL)}, a.failureTTL)
+	}
+
+	return user, ok, err
+}
+
+func (a *cachedTokenAuthenticator) getCachedResult(token string) (*cacheRecord, bool) {
+	record, ok := a.cache.get(token)
+	if !ok {
+		return nil, false
+	}
+	if record.expires.Before(a.clock.Now()) {
+		return nil, false
+	}
+	return record, true
+}
+
+func (a *cachedTokenAuthenticator) setCachedResult(key string, value *cacheRecord, ttl time.Duration) {
+	// Add to the cache
+	cachedValue := a.cache.set(key, value)
+
+	// If our object was the one stored, queue the removal
+	if cachedValue == value {
+		// TODO: batch removals to avoid a goroutine per cache item
```

Comment:
uhh.... this seems worse than the disease.  Have you measured against the next best solution: the existing LRU cache?